server/src/unifyfs_cmd_handler.c

```diff
@@ -479,18 +479,25 @@ static void unifyfs_mread_rpc(hg_handle_t handle)
                                in.bulk_handle, 0, bulk_handle, 0, size);
     assert(hret == HG_SUCCESS);
 
-    /* initiate read operations to fetch data for read requests */
-    int ret = rm_cmd_mread(in.app_id, in.client_id,
-                           in.read_count, buffer);
 
-    /* build our output values */
+    /* If we have successfully fetched requests, send the response first.
+     * Otherwise, the subsequent rm_cmd_mread() may try to send the data to the
+     * client, while the client is still waiting for our response.
+     * The read I/O results are reported via the shared memory (not through
+     * this rpc response).  */
     unifyfs_mread_out_t out;
-    out.ret = ret;
+    out.ret = 0;
```

Comment:
I haven't checked, but generally we want to return UNIFYFS_SUCCESS instead of 0.  In theory, UNIFYFS_SUCCESS might be redefined at some point to be some value other than 0.

However, the existing line was here to catch errors returned from the call to rm_cmd_mread to indicate to the caller that the request to the server failed to even get started, and so then the client knows that the server would not be sending any replies back at all.  Are we still accounting for that case?

Reply:
We could report those kind of "failed to launch" errors to the client through another mechanism.  Perhaps the client could always expect replies to come back, and have the server immediately send those replies with error status?

Reply:
@adammoody I see your point. Actually, I first tried with async rpcs (iforward/irespond) but later removed, because I thought the result of rm_cmd_mread was not too necessary. I will change back to the iforward, so we can also track the rm_cmd_mread.

Reply:
@adammoody I've resolved this by using iforward in client.